Adaptive Routing with Worker Load Based Strategy and Rule Based Strategy

YouNeedCryDear opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.ai to discuss further. -->
This PR introduces advanced routing capabilities to the SGLang router with two new policies and a unified routing context system that enables header-based worker selection.

## Motivation

<!-- Describe the purpose and goals of this pull request. -->

## Modifications
Key highlights:

  1. Two new routing policies:
    - RuleBasedPolicy - ordering by priority → cost → load with header filtering
    - LoadAwarePolicy - Simple minimum-load worker selection
  2. RoutingContext system:
    - Unified context bundling headers, request_text, and model_id
    - Enables header-aware routing decisions
  3. Router-level enhancements:
    - get_router_stats() method for multi-router scenarios
    - Enhanced select_router_for_request() using worker statistics
    - Priority, cost, and load-based router selection
  4. Request headers:
    - x-worker-priority - Minimum priority threshold
    - x-max-cost - Maximum cost threshold
    - x-prefer-pd - Prefer PD routers
<!-- Detail the changes made in this pull request. -->

Breaking changes:
// Before
  fn select_worker(&self, workers: &[Arc<dyn Worker>], request_text: Option<&str>) -> Option<usize>;

// After  
  fn select_worker(&self, workers: &[Arc<dyn Worker>], context: &RoutingContext) -> Option<usize>;

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->

## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

YouNeedCryDear committed:
fix test port collision

YouNeedCryDear committed:
add advanced routing logic for router

use advanced routing logic in router manager

use get candidate workers in both regular and PD router implementation

add unit test for advanced routing

add integration test for advanced routing

Revert "use advanced routing logic in router manager"

This reverts commit 03d58a8a827b74aa62574c7ac0fd73e01b5468d3.

revert all router-to-router strategy

remove router strategy test

slin1237 reviewed (commented):
im not entire following the design
can you write the description in the pr

YouNeedCryDear committed:
add load aware routing policy

add rule based routing policy

update reference for rule based and load aware policy

fix test case

remove integration test

add routingcontext including header request and change input for other policies

use context in load aware and rule based policy

add get router stats to router trait

update router selection logic based on router stats

add header to worker selection and get router stats function

format

YouNeedCryDear commented:
/gemini review

YouNeedCryDear committed:
remove liveness and readiness from router mod

YouNeedCryDear committed:
remove extra readiness from router

YouNeedCryDear committed:
format and check

YouNeedCryDear committed:
add load aware and rule based policy to cmd args for python and rust
